src/t8_data/t8_shmem.c

```diff
@@ -218,6 +218,201 @@ t8_shmem_array_allgather (const void *sendbuf, int sendcount,
                       recvcount, recvtype, recvarray->comm);
 }
 
+/**
+ * Computes the recvcounter- and displacementarray for 
+ * an MPI_Gatherv
+ * 
+ * \param[in] sendcount The number of items this proc sends
+ * \param[in, out] recvcounts On input a zero-initialized array that is going to be filled with the number of elements send by rank i
+ * \param[in, out] displs On input a zero-initialized  array that is going to be filled with the displacements
+ * \returns   The total number of items 
+ */
+static int
+t8_compute_recvcounts_displs (int sendcount, int *recvcounts, int *displs,
+                              int elem_size, sc_MPI_Comm comm)
+{
+  int                 mpisize;
+  int                 mpiret = sc_MPI_Comm_size (comm, &mpisize);
+  SC_CHECK_MPI (mpiret);
+
+  mpiret = sc_MPI_Allgather ((void *) &sendcount, 1, sc_MPI_INT, recvcounts,
+                             1, sc_MPI_INT, comm);
+  SC_CHECK_MPI (mpiret);
+
+  int                 recv_total = recvcounts[0];
+  for (int i = 1; i < mpisize; i++) {
+    displs[i] = displs[i - 1] + recvcounts[i - 1];
+    recv_total += recvcounts[i];
+  }
+
+  return recv_total;
+}
+
+#if defined(__bgq__) || defined(SC_ENABLE_MPIWINSHARED)
+/**
+ * Allgathers sendbuffer of various sizes and writes them in the recvarray. 
+ * Can be used for SC_SHMEM_WINDOW, SC_SHMEM_WINDOW_PRESCAN
+ * SC_SHMEM_BGQ or SC_SHMEMBGQ_PRESCAN (sc_shmem.h) 
+ * 
+ * \param[in] sendbuf   The sendbuffer
+ * \param[in] sendcount The number of items send by this proc
+ * \param[in] sendtype The type of items to send
+ * \param[in, out] recvarray The destination
+ * \param[in] recvtype The type of items to receive
+ * \param[in] comm The mpicommunicator to use. 
+ */
+static void
+t8_shmem_array_allgatherv_common (void *sendbuf,
+                                  const int sendcount,
+                                  sc_MPI_Datatype sendtype,
+                                  t8_shmem_array_t recvarray,
+                                  sc_MPI_Datatype recvtype,
+                                  sc_MPI_Comm comm, sc_MPI_Comm intranode,
+                                  sc_MPI_Comm internode)
+{
+  size_t              typesize;
+  int                 mpiret, intrarank, intrasize, intersize;
+  char               *noderecvchar = NULL;
+
+  typesize = sc_mpi_sizeof (recvtype);
+
+  mpiret = sc_MPI_Comm_rank (intranode, &intrarank);
+  SC_CHECK_MPI (mpiret);
+  mpiret = sc_MPI_Comm_size (intranode, &intrasize);
+  SC_CHECK_MPI (mpiret);
+  mpiret = sc_MPI_Comm_size (internode, &intersize);
+  SC_CHECK_MPI (mpiret);
+
+  /* intranode-gatherv */
+  int                *intra_displ = T8_ALLOC_ZERO (int, intrasize);
+  int                *intra_recvcounts = T8_ALLOC_ZERO (int, intrasize);
+  int                 intra_recv_total =
+    t8_compute_recvcounts_displs (sendcount, intra_recvcounts, intra_displ,
+                                  sizeof (sendtype), intranode);
+  if (!intrarank) {
```

Comment:
```suggestion
  if (intrarank == 0) {
```
I know its the same, but it makes it more obvoius, that this condition is only true for the intranode main rank